text/3594-expose-stackrealign-attribute.md

```diff
@@ -0,0 +1,85 @@
+- Feature Name: (`realign-stack-attr`)
+- Start Date: (2024-03-26)
+- RFC PR: [rust-lang/rfcs#3594](https://github.com/rust-lang/rfcs/pull/3594)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Expose the `alignstack` function attribute to rustc.
+
+# Motivation
+[motivation]: #motivation
+This is usefull when you have no guarantees about the alignment of the stack in an extern "C" function.
+Also verbatim from the attribute reference:
+Legacy x86 code uses 4-byte stack alignment. Newer aligned SSE instructions (like ‘movaps’) that work with the stack require operands to be 16-byte aligned. This attribute realigns the stack in the function prologue to make sure the stack can be used with SSE instructions.
+
+Interrupt service routines (ISRs) often require special treatment regarding stack alignment. When an interrupt occurs, the processor saves the current execution context onto the stack before transferring control to the ISR. However, the stack might not be aligned to the required boundary, especially in embedded systems where memory constraints are tight.
+
+By exposing a stack realignment feature to Rust, developers working on embedded systems or performance-critical applications gain the ability to guarantee stack alignment within ISRs and other critical code paths directly from Rust code. This not only simplifies development but also enhances the reliability and performance of systems that rely on interrupt handling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+The `[realign_stack]` attribute can be added to a function to force the compiler to add alignment to that function.
+Usefull in cases where your code is called from a thread or a binary compiled with another compiler, that uses different aligmnet and thus lead to a corruption.

```

Comment:
What is the usecase here? This sounds like the other compiler is just buggy. 

The motivation focuses on interrupts; if there's another use-case here then that should be introduced in the motivation, not here. Or it needs to be explained properly here.

Reply:
The other code the calls my functions was compiled in GCC with a stack alignment of 4.
Yet, I want to export my function to him. At this moment there is no "good" way to deal with this issue.

Reply:
That seems to contradict @Lokathor's statement that this is only about CPU-triggered interrupt calls.

Why is this not a blatant bug in the other code, outright violating the platform ABI? My understanding is that the alignment required by the C platform ABI is very clear.

Reply:
I have no control over other people's code though. I merely need to interact with it in a way that works and does not crash my process.

Reply:
Sorry, if there's other code that's clearly buggy then the default answer is not "let's add a new feature to Rust". Next time some code passes an argument in the wrong register, are we going to add a rustc flag to support that? No we're not, we're telling people to fix their code.

So we need to get to the bottom of this. Why do these people think it's okay to call C ABI functions with lower alignment? Is there ambiguity somewhere, or so? Something must be going on. Who even is the caller here? Can you link to a bug report where they said "sorry we won't fix this because X"? You need to provide a bit more than your vague statements so far if you want this to be relevant evidence.

Reply:
I am not sure exactly how they achieve this. I think they are compiled with a very old GCC as this is an old x86 platform.

Reply:
Can you give more details? What exactly is the target? What's the stack alignment requirement on that target? From what you said it almost sounds like that requirement changed at some point so old code would use a different requirement? But that sounds impossible, the C platform ABI can't have broken changes... or can it?

Please just explain properly what happens to someone who's never seen your code or anything like that code. that's the point of an RFC: to explain to the community the motivation for a language change. Don't make me tease out every single answer one by one. You didn't even answer half my questions...

Reply:
Ok, from what I gather about my target, It was compiled with an old GCC for a target that does not support SSE instructions.
"Modern Code" that supports SSE, aligns the stack to 16. However because my target does not support SSE it aligns its stack to 4 (that's the way it was compiled and from what I understood `-mpreferred-stack-boundary=2` was used).
The proposed stack alignment attribute will allow us mix this legacy code with new code that does keep it's stack to 16 bytes alignment and that's what `force_align_arg_pointer` in GCC is used for.

Reply:
Thanks.

So what is that target, something like `i686-unknown-linux-gnu`? But that sounds like the target changed its C platform ABI, which is mind-boggling? Like, does this mean that if I try to mix old and new libraries for this target things will go wrong as at some point the target just changed the alignment that is generally required for functions...?

Reply:
tbh i wouldn't be surprised if some particularly old x86 target changed the ABI, after all they kinda started with a yolo ABI

Reply:
> So what is that target, something like `i686-unknown-linux-gnu`?

At the start we used that target.
Now we use `i586-unknown-linux-gnu` because it is the closer to the target.

And yeah it took us like a lot of time to understand the issue was a stack alignment issue. 
At random places it stopped working and adding traces to understand where the code crashed just caused the code to work again. 
Hell of a time.

Reply:
Yeah that does sound painful.

It doesn't answer my question though about where this issue comes from. Is `-mpreferred-stack-boundary` just a wildly unsafe compiler flag that produces programs with a different ABI...? That sounds like an outrageously dangerous flag to provide? Or is there some sane reason for why we got where we got?

Reply:
For the record, GCC has a number of compiler flags that change the ABI incompatibly.  If you browse the [list of machine-dependent options](https://gcc.gnu.org/onlinedocs/gcc/Submodel-Options.html), you’ll find a bunch.

Reply:
@benisxdxd thanks, that helps explain things. Would be good to include a summary of this as rationale in the RFC. It *seems* like the exact same thing (making functions "robust" to insufficiently aligned stacks) works for both your use-case and the interrupt use-case, which is good. So the RFC should say: we want such "robust" functions, and here are two use-cases where they come up and why there's not really a good alternative for either of them.